pkg/ioutils/readers.go

```diff
@@ -14,6 +14,34 @@ import (
 
 var rndSrc = random.NewSource()
 
+// ReadCloserWithError is used to wrap a reader but also be able to close the
+// reader given a specific error. Successive calls to Read() will return the
+// given error.
+type ReadCloserWithError struct {
+	io.Reader
+	err error
```

Comment:
You could have the same functionality with a buffered channel, but have confidence that concurrent reads/writes won't cause crashes. For example:

In the stuct:

    errc chan error

In NewReadCloserWithError:

    return &ReadCloserWithError{Reader: r, errc: make(chan error, 1)}

In Read:

    select {
    case err := <-rce.errc:
        return 0, err
    }

In CloseWithError:

    select {
    case rce.errc <- err:
    }